include/dslash_helper.cuh

```diff
@@ -4,17 +4,34 @@
 #include <gauge_field.h>
 #include <register_traits.h>
 #include <index_helper.cuh>
+#include <shmem_helper.cuh>
+
+#ifdef NVSHMEM_COMMS
+namespace
+{
+  __device__ cuda::atomic<long, cuda::thread_scope_system> intersync {0};
+  __device__ cuda::atomic<long, cuda::thread_scope_device> interior_done {0};
+  __device__ cuda::atomic<int, cuda::thread_scope_block> interior_count {0};
+} // namespace
```

Comment:
Since we require `cuda::atomic`, I think we need to exclude `SANITIZE` builds like we do with heterogeneous atomics, since libcu++ doesn't play nicely with ASAN.

Reply:
Another thought: since these are statically declared per compilation unit, we end up with this allocation overhead repeated for every dslash file.  Is there any reason not to allocate these dynamically, say in dslash_quda.cu, and reuse these atomics for all stencils to avoid this overhead?  Initialization could be done with a placement new constructor similar to the atomics used with the heterogeneous atomic reductions.

Reply:
I blocked the SANITIZE build for now.

Reply:
Best we can do for now until libcu++ fixes this

Reply:
I'll look into placement new for that. I guess the same applies to the ones in `dslash_pack.cuh`. Should we put that in now or for future optimization? I don't have a feeling for the actual overhead here.

Reply:
Looking at the output from nvcc, the overhead is 64 KiB per translation unit I think.  It should be fairly easy to make it a single instance with placement new, so my preference would be to do this now.

Reply:
Just commit this. Please have a look.

Reply:
Ok, this looks mostly good now.  I would though prefer having a getter method rather than having the variables extern, and made global.  We've got enough globals in QUDA 😉